File: ts/src/Token.ts
import { Buffer } from "buffer"
import { sha256 } from "js-sha256"

export class Token {
  /**
   * The Semifinals Authentication Token epoch time (1st January 2023) in seconds.
   */
  public static epoch: number = Date.UTC(2023, 0, 1) / 1000

  /**
   * The separator used to partition the token.
   */
  public static separator: string = "."

  /**
   * Generate a token for the given ID.
   * @param id The ID for the token
   * @param secret The secret used for the sha256 algorithm
   * @returns The generated token
   */
  public static generate(id: string, secret: string): string {
    const timestamp = Date.now()

    const payload = Token.payload(id, timestamp)
    const signature = Token.sign(payload, secret)

    return payload + Token.separator + signature
  }

  /**
   * Generate a payload using the given ID and timestamp.
   * @param id The ID for the token
   * @param timestamp The unix timestamp for the token
   * @returns The resulting payload
   */
  public static payload(id: string, timestamp: number): string {
    const encodedId = Token.toBase64(id)
    const encodedTimestamp = Token.toBase64(
      String(Math.floor(timestamp / 1000) - Token.epoch)
    )

    return encodedId + Token.separator + encodedTimestamp
  }

  /**
   * Generate a signature for the given payload.
   * @param payload The payload to sign
   * @param secret The secret used for the sha256 algorithm
   * @returns The generated signature
   */
  public static sign(payload: string, secret: string): string {
    const hash = sha256.hmac.update(secret, payload).digest()
    return Buffer.from(hash).toString("base64")
  }

  /**
   * Validate that the token looks correct in structure.
   * @param token The token to validate
   * @returns Whether or not the token is valid
   */
  public static validate(token: string): boolean {
    const parts = Token.split(token)
    if (parts.length !== 3) return false

    const id = Token.toUtf8(parts[0])
    if (id.length === 0) return false

    const unparsedTimestamp = Token.toUtf8(parts[1])
    if (unparsedTimestamp.length === 0) return false

    const timestamp = parseInt(unparsedTimestamp)
    if (typeof timestamp !== "number") return false

    const signature = parts[2]
    if (signature.length === 0) return false

    return true
  }

  /**
   * Verify that the signature of the token is authentic.
   * @param token The full token
   * @param secret The secret used for the sha256 algorithm
   * @returns Whether or not the token was verified to be authentic
   */
  public static verify(token: string, secret: string): boolean {
    if (!Token.validate(token)) return false

    const id = Token.getId(token)
    const timestamp = Token.getTimestamp(token)
    const signature = Token.getSignature(token)

    if (id === null || timestamp === null || signature === null) return false

    const payload = Token.payload(id, timestamp)
    const generatedSignature = Token.sign(payload, secret)

    return signature === generatedSignature
  }

  /**
   * Get the ID of the token.
   * @param token The full token
   * @returns The ID of the token
   */
  public static getId(token: string): string | null {
    if (!Token.validate(token)) return null

    const parts = Token.split(token)
    const id = Token.toUtf8(parts[0])

    return id
  }

  /**
   * Get the unix timestamp of the token.
   * @param token The full token
   * @returns The unix timestamp of the token in milliseconds
   */
  public static getTimestamp(token: string): number | null {
    if (!Token.validate(token)) return null

    const parts = Token.split(token)
    const timestamp = parseInt(Token.toUtf8(parts[1]))

    return (Token.epoch + timestamp) * 1000
  }

  /**
   * Get the signature from a token.
   * @param token The full token
   * @returns The token's signature
   */
  public static getSignature(token: string): string | null {
    if (!Token.validate(token)) return null

    const parts = Token.split(token)
    const signature = parts[2]

    return signature
  }

  /**
   * Split a string into parts by the token separator.
   * @param str The string to split
   * @returns The string split into its parts
   */
  public static split(str: string): string[] {
    return str.split(Token.separator)
  }

  /**
   * Convert a utf8 string into base64.
   * @param str The utf8 string
   * @returns The string converted to base64
   */
  public static toBase64(str: string): string {
    return Buffer.from(str, "utf8").toString("base64")
  }

  /**
   * Convert a base64 string into utf8.
   * @param str The base64 string
   * @returns The string converted to utf8
   */
  public static toUtf8(str: string): string {
    return Buffer.from(str, "base64").toString("utf8")
  }
}

File: cs/Token/Token.cs
using System.Security.Cryptography;
using System.Text;

namespace Semifinals.Utils.Tokens;

public class Token
{
    /// <summary>
    /// The Semifinals Authentication Token epoch time (1st January 2023) in seconds.
    /// </summary>
    public static readonly long Epoch = ((DateTimeOffset)new DateTime(2023, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc)).ToUnixTimeSeconds();

    /// <summary>
    /// The separator used to partition the token.
    /// </summary>
    public static readonly char Separator = '.';

    /// <summary>
    /// Generate a token for the given ID.
    /// </summary>
    /// <param name="id">The ID for the token</param>
    /// <param name="secret">The secret used for the sha256 algorithm</param>
    /// <returns>The generated token</returns>
    public static string Generate(string id, string secret)
    {
        long timestamp = ((DateTimeOffset)DateTime.UtcNow).ToUnixTimeMilliseconds();

        string payload = Payload(id, timestamp);
        string signature = Sign(payload, secret);

        return payload + Separator + signature;
    }

    /// <summary>
    /// Generate a payload using the given ID and timestamp.
    /// </summary>
    /// <param name="id">The ID for the token</param>
    /// <param name="timestamp">The timestamp for the token</param>
    /// <returns>The resulting payload</returns>
    public static string Payload(string id, long timestamp)
    {
        string encodedId = ToBase64(id);
        string encodedTimestamp = ToBase64((timestamp / 1000 - Epoch).ToString());

        return encodedId + Separator + encodedTimestamp;
    }

    /// <summary>
    /// Generate a signature for the given payload.
    /// </summary>
    /// <param name="payload">The payload to sign</param>
    /// <param name="secret">The secret used for the sha256 algorithm</param>
    /// <returns>The generated signature</returns>
    public static string Sign(string payload, string secret)
    {
        byte[] payloadBytes = Encoding.UTF8.GetBytes(payload);

        byte[] secretBytes = Encoding.UTF8.GetBytes(secret);

        using HMACSHA256 hmac = new(secretBytes);
        byte[] signatureBytes = hmac.ComputeHash(payloadBytes);

        string signature = Convert.ToBase64String(signatureBytes);
        return signature;
    }

    /// <summary>
    /// Validate that the token looks correct in structure.
    /// </summary>
    /// <param name="token">The token to validate</param>
    /// <returns>Whether or not the token is valid</returns>
    public static bool Validate(string token)
    {
        string[] parts = Split(token);
        if (parts.Length != 3)
            return false;

        string id = ToUtf8(parts[0]);
        if (id.Length == 0)
            return false;

        string unparsedTimestamp = ToUtf8(parts[1]);
        if (unparsedTimestamp.Length == 0)
            return false;

        if (!long.TryParse(unparsedTimestamp, out long _))
            return false;

        string signature = parts[2];
        if (signature.Length == 0)
            return false;

        return true;
    }

    /// <summary>
    /// Verify that the signature of the token is authentic.
    /// </summary>
    /// <param name="token">The full token</param>
    /// <param name="secret">The secret used for the sha256 algorithm</param>
    /// <returns>Whether or not the token was verified to be authentic</returns>
    public static bool Verify(string token, string secret)
    {
        if (!Validate(token))
            return false;

        string? id = GetId(token);
        long? timestamp = GetTimestamp(token);
        string? signature = GetSignature(token);

        if (id is null || timestamp is null || signature is null)
            return false;

        string payload = Payload(id, (long)timestamp);
        string generatedSignature = Sign(payload, secret);

        return signature == generatedSignature;
    }

    /// <summary>
    /// Get the ID of the token.
    /// </summary>
    /// <param name="token">The full token</param>
    /// <returns>The ID of the token</returns>
    public static string? GetId(string token)
    {
        if (!Validate(token))
            return null;

        string[] parts = Split(token);
        string id = ToUtf8(parts[0]);

        return id;
    }

    /// <summary>
    /// Get the unix epoch timestamp of the token.
    /// </summary>
    /// <param name="token">The full token</param>
    /// <returns>The Semifinals epoch timestamp of the token in milliseconds</returns>
    public static long? GetTimestamp(string token)
    {
        if (!Validate(token))
            return null;

        string[] parts = Split(token);
        
        if (!long.TryParse(ToUtf8(parts[1]), out long timestamp))
            return null;
            

        return (Epoch + timestamp) * 1000;
    }

    /// <summary>
    /// Get the signature from a token.
    /// </summary>
    /// <param name="token">The full token</param>
    /// <returns>The token's signature</returns>
    public static string? GetSignature(string token)
    {
        if (!Validate(token))
            return null;

        string[] parts = Split(token);
        string signature = parts[2];

        return signature;
    }

    /// <summary>
    /// Split a string into parts by the token separator.
    /// </summary>
    /// <param name="str">The string to split</param>
    /// <returns>The string split into its parts</returns>
    public static string[] Split(string str)
    {
        return str.Split(Separator);
    }

    /// <summary>
    /// Convert a utf8 string into base64
    /// </summary>
    /// <param name="str">The utf8 string</param>
    /// <returns>The string converted to base64</returns>
    public static string ToBase64(string str)
    {
        try
        {
            byte[] bytes = Encoding.UTF8.GetBytes(str);
            return Convert.ToBase64String(bytes);
        }
        catch (FormatException)
        {
            return "";
        }
    }

    /// <summary>
    /// Convert a base64 string into utf8.
    /// </summary>
    /// <param name="str">The base64 string</param>
    /// <returns>The string converted to utf8</returns>
    public static string ToUtf8(string str)
    {
        try
        {
            byte[] bytes = Convert.FromBase64String(str);
            return Encoding.UTF8.GetString(bytes);
        }
        catch (FormatException)
        {
            return "";
        }
    }
}